gateway/Dockerfile

```diff
@@ -1,37 +1,61 @@
-FROM registry.access.redhat.com/ubi9-minimal:9.4@sha256:104cf11d890aeb7dd5728b7d7732e175a0e4018f1bb00d2faebcc8f6bf29bd52
-RUN microdnf install -y python3.11-3.11.7 python3.11-pip-22.3.1 python3.11-devel-3.11.7 vim-enhanced-8.2.2637 &&\
-    microdnf clean all
-RUN ln -s /usr/bin/python3.11 /usr/local/bin/python3 && \
-    ln -s /usr/bin/python3.11 /usr/local/bin/python &&\
-    ln -s /usr/bin/pip3.11 /usr/local/bin/pip3 &&\
-    ln -s /usr/bin/pip3.11 /usr/local/bin/pip
+ARG MICRO_IMAGE_DIR=/ubi-micro-img
 
+# BASE image using UBI 9 micro where the
+# application and requirements will be installed
+FROM registry.access.redhat.com/ubi9-micro:9.4-15 AS BASE
+
+# BUILD image using UBI 9 where the dependencies that
+# require installing with a package manager will be installed
+FROM registry.access.redhat.com/ubi9:9.4-1214.1726694543 AS BUILD
+ARG MICRO_IMAGE_DIR
+
+# Copy the BASE image into the BUILD image
+RUN mkdir ${MICRO_IMAGE_DIR}
+COPY --from=BASE / ${MICRO_IMAGE_DIR}
+
+# Install Python inside the BASE image
+# hadolint ignore=DL3041
+RUN dnf install --installroot ${MICRO_IMAGE_DIR} --nodocs -y \
+    python3.11-3.11.7 \
+    python3.11-devel-3.11.7 \
+    libstdc++ &&\
+    dnf upgrade --installroot ${MICRO_IMAGE_DIR} --nodocs -y && \
+    dnf clean all --installroot ${MICRO_IMAGE_DIR}
+
+# APP image from `scratch` which will be the final image
+# and remaining application requirements will be installed
+FROM scratch AS APP
```

Comment:
I wonder why this APP image is necessary.  What is the difference between copying back the ${MICRO_IMAGE_DIR}/ into the BASE image and the APP image?

Reply:
The BASE image still has things like a package manager and a shell that the gateway really doesn't need, whereas scratch / APP is empty except for what we explicitly add to it.  So that does two things: (a) it reduces the number of components that we have to manage for CVEs, and (b) it reduces the threat vectors for that particular container.

Reply:
I may be wrong but I read 
```
COPY --from=BASE / ${MICRO_IMAGE_DIR}
```
copies everything in the BASE to ${MICRO_IMAGE_DIR} directory and
```
COPY --from=BUILD ${MICRO_IMAGE_DIR}/ .
```
copies everything in the BASE + additional pieces to the APP
So everything in the BASE is copied to APP.  Where am I misunderstanding?

Reply:
You could be right... I'm just copying from what [another team already did](https://github.ibm.com/IBM-Q-Software/qiskit-code-assistant-service/blob/7fd8d83b161c5c67d15c3159f41e2be51638794e/Dockerfile) and didn't look too closely at what they were doing, I just assumed it worked since it had already gone through review.